README.md
picuplib
=========

Small python library for uploading pictures to [picflash.org](https://picflash.org)

Requires python 2.6 or better.
Python 3 is supported too.

Documentation can be found on [ReadTheDocs](http://picuplib.readthedocs.org/en/latest/)

Installation
============
1. Download and unpack the tarball or clone the repo
2. type ``python setup.py install``


CHANGELOG
v0.2.4 - 2014-09-27
- Fixed json output

v0.2.3 - 2014-09-11
-Implemented workaround for piclfash unicode problems

v0.2.1 - 2014-09-09
-Added setup.py pypi classefiers


picuplib/checks.py
# -*- coding:utf8 -*-
######################## BEGIN LICENSE BLOCK ########################
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA
######################### END LICENSE BLOCK #########################
"""
module for some argument cheking
"""

from json import loads

from picuplib.exceptions import (UnsuportedResize, UnsuportedRotation,
                                 UnsupportedFormat, UnkownError)
from picuplib.globals import ALLOWED_ROTATION, ALLOWED_SIZE


def check_rotation(rotation):
    """checks rotation parameter if illegal value raises exception"""

    if rotation not in ALLOWED_ROTATION:
        allowed_rotation = ', '.join(ALLOWED_ROTATION)
        raise UnsuportedRotation('Rotation %s is not allwoed. Allowed are %s'
                                 % (rotation, allowed_rotation))

def check_size(size):
    """checks size parameter if illegal value raises exception"""

    if size not in ALLOWED_SIZE:
        allowed_size = ', '.join(ALLOWED_SIZE)
        raise UnsuportedResize('Size %s is not allowed. Allowed are %s'
                               % (size, allowed_size))

def check_noexif(noexif):
    """checks if noexif parameter is boolean"""
    if not isinstance(noexif, bool):
        raise TypeError('noexif must be boolean')

def check_response(response):
    """
    checks the response if the server returned an error raises an exception.
    """
    response = response.replace('null', '')
    response = loads(response)
    if 'failure' in response:
        if response['failure'] == 'Falscher Dateityp':
            raise UnsupportedFormat('Please look at picflash.org '
                                    'witch formats are supported')
        else:
            raise UnkownError(response['failure'])


picuplib/globals.py
# -*- coding:utf8 -*-
######################## BEGIN LICENSE BLOCK ########################
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA
######################### END LICENSE BLOCK #########################
"""
Module for some "global" constants
"""

API_URL = 'https://picflash.org/tool.php'

ALLOWED_SIZE = ('80x80', '100x75', '100x100', '150x112', '468x60', '400x400',
                '320x240', '640x480', '800x600', '1024x768', '1280x1024',
                '1600x1200', 'og')

ALLOWED_ROTATION = ('00', '90', '180', '270')




picuplib/__init__.py
# -*- coding:utf8 -*-
######################## BEGIN LICENSE BLOCK ########################
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA
######################### END LICENSE BLOCK #########################
"""
Small picflash upload library
"""

__version__ = '0.2'

from .upload import upload, remote_upload, Upload




picuplib/exceptions.py
# -*- coding: utf8 -*-
######################## BEGIN LICENSE BLOCK ########################
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA
######################### END LICENSE BLOCK #########################

"""
Collections of custom exceptions
"""

from __future__ import unicode_literals, print_function


class UnsuportedRotation(Exception):
    """
    Raised when unsupported rotation value is given.
    """
    pass

class UnsuportedResize(Exception):
    """
    Raised when unsupported resize value is given
    """
    pass

class UnsupportedFormat(Exception):
    """
    Raised when file format is unsupported
    """
    pass


picuplib/upload.py
# -*- coding:utf8 -*-
######################## BEGIN LICENSE BLOCK ########################
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301 USA
######################### END LICENSE BLOCK #########################
"""
This module handels the entire upload and some argument and response checking
"""

from __future__ import unicode_literals, print_function

from requests import post

from picuplib.checks import (check_resize, check_rotation, check_noexif,
                             check_response)

from picuplib.globals import API_URL

class Upload(object):
    """
    Class based wrapper for uploading.
    It stores the apikey and default settings for resize, rotation …

    :param str apikey: Apikey needed for Autentication on picflash.
    :param str resize: Either 'og' or an allowed resolution in the followong\
        format '80x80'
    :param (str|degree) rotation: The picture will be rotated by this Value. \
        Allowed values are 00, 90, 180, 270.
    :param boolean noexif: set to True when exif data should be purged.
    """

    def __init__(self, apikey, resize='og', rotation='00', noexif=False):
        self._apikey = apikey
        self._resize = resize
        self._rotation = rotation
        self._noexif = noexif

    @property
    def resize(self):
        """getter for _resize"""
        return self._resize

    @resize.setter
    def resize(self, value):
        """setter for _resize"""
        check_resize(value)
        self._resize = value

    @property
    def rotation(self):
        """getter for _rotation"""
        return self._rotation

    @rotation.setter
    def rotation(self, value):
        """setter for rotation"""
        check_rotation(value)
        self._rotation = value

    @property
    def noexif(self):
        """getter for _noexif"""
        return self._noexif

    @noexif.setter
    def noexif(self, value):
        """setter for _noexif"""
        check_noexif(value)
        self._noexif = value

    def upload(self, picture, resize=None, rotation=None, noexif=None):
        """wraps upload function"""
        if not resize:
            resize = self._resize
        if not rotation:
            rotation = self._rotation
        if not noexif:
            noexif = self._noexif

        return upload(self._apikey, picture, resize, rotation, noexif)

    def remote_upload(self, picture_url, resize=None,
                      rotation=None, noexif=None):
        """wraps remote_upload funktion"""
        if not resize:
            resize = self._resize
        if not rotation:
            rotation = self._rotation
        if not noexif:
            noexif = self._noexif

        return remote_upload(self._apikey, picture_url,
                             resize, rotation, noexif)


def upload(apikey, picture, resize='og', rotation='00', noexif=False):
    """
    prepares post for regular upload

    param: str apikey
    """
    check_rotation(rotation)
    check_resize(resize)

    post_data = compose_post(apikey, resize, rotation, noexif)

    with open(picture, 'rb') as file_obj:
        post_data['Datei[]'] = file_obj

        return do_upload(post_data)

def remote_upload(apikey, picture_url, resize='og',
                  rotation='00', noexif=False):
    """
    prepares post for remote upload
    """
    check_rotation(rotation)
    check_resize(resize)

    post_data = compose_post(apikey, resize, rotation, noexif)
    post_data['url[]'] = ('', picture_url)

    return do_upload(post_data)



def compose_post(apikey, resize, rotation, noexif):
    """
    composes basic post requests
    """
    check_rotation(rotation)
    check_resize(resize)

    post_data = {
        'formatliste': ('', resize),
        'userdrehung': ('', rotation),
        'apikey': ('', apikey)
        }

    if noexif:
        post_data['noexif'] = ('', '')

    return post_data

def do_upload(post_data):
    """
    does the actual upload
    """
    response = post(API_URL, files=post_data)
    check_response(response.text)

    return response.json()

